File: .verify/VersionLogic.java
/* Replica exata da logica corrigida em Tools.java (AUDITORIA 5.2 e 5.3)
 * e da antiga, para comparar comportamento lado a lado. */
public class VersionLogic {

    /* ---------- implementacao NOVA (corrigida) ---------- */
    static int versionComponent(String[] components, int index) {
        if (components == null || index >= components.length) return -1;
        String raw = components[index].trim();
        int end = 0;
        while (end < raw.length() && Character.isDigit(raw.charAt(end))) end++;
        if (end == 0) return -1;
        try { return Integer.parseInt(raw.substring(0, end)); }
        catch (NumberFormatException e) { return -1; }
    }

    static boolean versionAtLeast(String[] components, int minMajor, int minMinor) {
        int major = versionComponent(components, 0);
        if (major < 0) return false;
        if (major != minMajor) return major > minMajor;
        int minor = versionComponent(components, 1);
        if (minor < 0) return false;
        return minor >= minMinor;
    }

    /* ---------- implementacao ANTIGA (com bugs) ---------- */
    static String oldBehaviour(String libName) {
        try {
            String[] version = libName.split(":")[2].split("\\.");
            boolean skip = Integer.parseInt(version[0]) >= 5 && Integer.parseInt(version[1]) >= 13;
            return skip ? "mantem" : "troca p/ 5.13.0";
        } catch (ArrayIndexOutOfBoundsException e) {
            return "CRASH: ArrayIndexOutOfBounds";
        } catch (NumberFormatException e) {
            return "CRASH: NumberFormatException";
        }
    }

    static String newBehaviour(String libName) {
        if (libName == null) return "ignorado";
        String[] nameParts = libName.split(":");
        if (nameParts.length < 3) return "ignorado (coordenada malformada)";
        String[] version = nameParts[2].split("\\.");
        return versionAtLeast(version, 5, 13) ? "mantem" : "troca p/ 5.13.0";
    }

    public static void main(String[] args) {
        String[] cases = {
            "net.java.dev.jna:jna:5.13.0",          // novo o suficiente -> mantem
            "net.java.dev.jna:jna:5.14.0",          // mais novo -> mantem
            "net.java.dev.jna:jna:6.2.1",           // BUG 5.3: antigo fazia DOWNGRADE
            "net.java.dev.jna:jna:6.0.0",           // BUG 5.3: idem
            "net.java.dev.jna:jna:4.5.2",           // antigo de verdade -> troca
            "net.java.dev.jna:jna:5.12.9",          // minor menor -> troca
            "net.java.dev.jna:jna:5.13.0-SNAPSHOT", // BUG 5.2: NumberFormatException
            "net.java.dev.jna:jna:5",               // BUG 5.2: AIOOBE (sem minor)
            "net.java.dev.jna:jna:2.0.0-beta",      // BUG 5.2: NumberFormatException
            "net.java.dev.jna:jna:1.0.0+build1",    // BUG 5.2: NumberFormatException
            "grupo:artefato",                       // BUG 5.2: AIOOBE (sem versao)
        };

        System.out.printf("%-38s | %-28s | %s%n", "LIBRARY", "ANTIGO", "NOVO");
        System.out.println("-".repeat(100));
        int crashes = 0, downgrades = 0;
        for (String c : cases) {
            String o = oldBehaviour(c), n = newBehaviour(c);
            if (o.startsWith("CRASH")) crashes++;
            if (o.equals("troca p/ 5.13.0") && n.equals("mantem")) downgrades++;
            System.out.printf("%-38s | %-28s | %s%n", c, o, n);
        }
        System.out.println("-".repeat(100));
        System.out.println("Crashes evitados: " + crashes);
        System.out.println("Downgrades indevidos evitados: " + downgrades);
        if (crashes == 0 || downgrades == 0) {
            System.out.println("FALHA: o teste deveria demonstrar ambos os problemas");
            System.exit(1);
        }
        System.out.println("OK");
    }
}

File: .verify/version_logic.py
"""Port 1:1 da logica corrigida em Tools.java (AUDITORIA 5.2 e 5.3),
comparada com a implementacao antiga, para provar que os crashes e o
downgrade indevido deixam de ocorrer."""


# ---------- implementacao NOVA (espelha o Java) ----------
def version_component(components, index):
    if components is None or index >= len(components):
        return -1
    raw = components[index].strip()
    end = 0
    while end < len(raw) and raw[end].isdigit():
        end += 1
    if end == 0:
        return -1
    return int(raw[:end])


def version_at_least(components, min_major, min_minor):
    major = version_component(components, 0)
    if major < 0:
        return False
    if major != min_major:
        return major > min_major
    minor = version_component(components, 1)
    if minor < 0:
        return False
    return minor >= min_minor


# ---------- implementacao ANTIGA (com bugs) ----------
def old_behaviour(lib_name):
    try:
        version = lib_name.split(":")[2].split(".")
        skip = int(version[0]) >= 5 and int(version[1]) >= 13
        return "mantem" if skip else "troca p/ 5.13.0"
    except IndexError:
        return "CRASH: ArrayIndexOutOfBounds"
    except ValueError:
        return "CRASH: NumberFormatException"


def new_behaviour(lib_name):
    if lib_name is None:
        return "ignorado"
    name_parts = lib_name.split(":")
    if len(name_parts) < 3:
        return "ignorado (coordenada malformada)"
    version = name_parts[2].split(".")
    return "mantem" if version_at_least(version, 5, 13) else "troca p/ 5.13.0"


CASES = [
    ("net.java.dev.jna:jna:5.13.0", "mantem"),
    ("net.java.dev.jna:jna:5.14.0", "mantem"),
    ("net.java.dev.jna:jna:6.2.1", "mantem"),            # 5.3: antigo fazia downgrade
    ("net.java.dev.jna:jna:6.0.0", "mantem"),            # 5.3: idem
    ("net.java.dev.jna:jna:4.5.2", "troca p/ 5.13.0"),
    ("net.java.dev.jna:jna:5.12.9", "troca p/ 5.13.0"),
    ("net.java.dev.jna:jna:5.13.0-SNAPSHOT", "mantem"),  # 5.2
    ("net.java.dev.jna:jna:5", "troca p/ 5.13.0"),       # 5.2
    ("net.java.dev.jna:jna:2.0.0-beta", "troca p/ 5.13.0"),
    ("net.java.dev.jna:jna:1.0.0+build1", "troca p/ 5.13.0"),
    ("grupo:artefato", "ignorado (coordenada malformada)"),
]


def main():
    print(f"{'LIBRARY':<38} | {'ANTIGO':<28} | {'NOVO':<32} | ESPERADO")
    print("-" * 125)
    crashes = downgrades = failures = 0
    for lib, expected in CASES:
        old = old_behaviour(lib)
        new = new_behaviour(lib)
        if old.startswith("CRASH"):
            crashes += 1
        if old == "troca p/ 5.13.0" and new == "mantem":
            downgrades += 1
        ok = new == expected
        if not ok:
            failures += 1
        mark = "" if ok else "   <-- FALHOU"
        print(f"{lib:<38} | {old:<28} | {new:<32} | {expected}{mark}")
    print("-" * 125)
    print(f"Crashes evitados pela correcao 5.2 : {crashes}")
    print(f"Downgrades indevidos evitados (5.3): {downgrades}")
    print(f"Divergencias do esperado           : {failures}")
    assert failures == 0, "A nova implementacao divergiu do comportamento esperado"
    assert crashes > 0 and downgrades > 0, "O teste deveria demonstrar ambos os problemas"
    print("\nOK - todas as assercoes passaram")


if __name__ == "__main__":
    main()
